File: problems/3529-count-cells-in-overlapping-horizontal-and-vertical-substrings/solution.py
class Solution:
    def countCells(self, grid: List[List[str]], pattern: str) -> int:
        m = len(grid)
        n = len(grid[0])
        p_len = len(pattern)
        horz_matches = set()
        vert_matches = set()
        count = 0

        for r in range(m):
            for c in range(n):
                # Check horizontal
                horz_str = ""
                curr_r = r
                curr_c = c
                for _ in range(p_len):
                    horz_str += grid[curr_r][curr_c]
                    curr_c += 1
                    if curr_c == n:
                        curr_c = 0
                        curr_r += 1
                        if curr_r == m:
                            break

                if len(horz_str) == p_len and horz_str == pattern:
                    curr_r = r
                    curr_c = c
                    for _ in range(p_len):
                        horz_matches.add((curr_r, curr_c))
                        curr_c += 1
                        if curr_c == n:
                            curr_c = 0
                            curr_r += 1

        for c in range(n):
            for r in range(m):
                # Check vertical
                vert_str = ""
                curr_r = r
                curr_c = c
                for _ in range(p_len):
                    vert_str += grid[curr_r][curr_c]
                    curr_r += 1
                    if curr_r == m:
                        curr_r = 0
                        curr_c += 1
                        if curr_c == n:
                            break
                            
                if len(vert_str) == p_len and vert_str == pattern:
                    curr_r = r
                    curr_c = c
                    for _ in range(p_len):
                        vert_matches.add((curr_r, curr_c))
                        curr_r += 1
                        if curr_r == m:
                            curr_r = 0
                            curr_c += 1

        for cell in horz_matches:
            if cell in vert_matches:
                count += 1

        return count

File: problems/3529-count-cells-in-overlapping-horizontal-and-vertical-substrings/solution.java
class Solution {
    public int countCells(char[][] grid, String pattern) {
        int m = grid.length;
        int n = grid[0].length;
        int patternLength = pattern.length();
        boolean[][] horizontal = new boolean[m][n];
        boolean[][] vertical = new boolean[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int k = 0;
                while (k < patternLength) {
                    int currentRow = i + (j + k) / n;
                    int currentCol = (j + k) % n;
                    if (currentRow >= m || grid[currentRow][currentCol] != pattern.charAt(k)) {
                        break;
                    }
                    k++;
                }
                if (k == patternLength) {
                    for (int l = 0; l < patternLength; l++) {
                        int currentRow = i + (j + l) / n;
                        int currentCol = (j + l) % n;
                        horizontal[currentRow][currentCol] = true;
                    }
                }
            }
        }

        for (int j = 0; j < n; j++) {
            for (int i = 0; i < m; i++) {
                int k = 0;
                while (k < patternLength) {
                    int currentCol = j + (i + k) / m;
                    int currentRow = (i + k) % m;
                    if (currentCol >= n || grid[currentRow][currentCol] != pattern.charAt(k)) {
                        break;
                    }
                    k++;
                }
                if (k == patternLength) {
                    for (int l = 0; l < patternLength; l++) {
                        int currentCol = j + (i + l) / m;
                        int currentRow = (i + l) % m;
                        vertical[currentRow][currentCol] = true;
                    }
                }
            }
        }

        int count = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (horizontal[i][j] && vertical[i][j]) {
                    count++;
                }
            }
        }

        return count;
    }
}